pinot-core/src/main/java/org/apache/pinot/core/operator/filter/H3IndexFilterOperator.java

```diff
@@ -50,94 +46,164 @@
+      coordinate = GeometrySerializer.deserialize(BytesUtils.toBytes(arguments.get(0).getLiteral())).getCoordinate();
     }
-    _dataSource = indexSegment.getDataSource(columnName);
-    _h3IndexReader = _dataSource.getH3Index();
-    _resolution = _h3IndexReader.getH3IndexResolution();
-    Preconditions.checkArgument(predicate instanceof RangePredicate,
-        String.format("H3 index does not support predicate type %s", predicate.getType()));
+    assert _h3IndexReader != null;
+    int resolution = _h3IndexReader.getH3IndexResolution().getLowestResolution();
+    _h3Id = H3Utils.H3_CORE.geoToH3(coordinate.y, coordinate.x, resolution);
+    _edgeLength = H3Utils.H3_CORE.edgeLength(resolution, LengthUnit.m);
+
     RangePredicate rangePredicate = (RangePredicate) predicate;
-    _distance = Double.parseDouble(rangePredicate.getUpperBound());
-    _numDocs = numDocs;
-    try {
-      _h3Core = H3Core.newInstance();
-    } catch (IOException e) {
-      throw new RuntimeException("Unable to instantiate H3 instance", e);
+    if (!rangePredicate.getLowerBound().equals(RangePredicate.UNBOUNDED)) {
+      _lowerBound = Double.parseDouble(rangePredicate.getLowerBound());
+    } else {
+      _lowerBound = Double.NaN;
+    }
+    if (!rangePredicate.getUpperBound().equals(RangePredicate.UNBOUNDED)) {
+      _upperBound = Double.parseDouble(rangePredicate.getUpperBound());
+    } else {
+      _upperBound = Double.NaN;
     }
   }
 
   @Override
   protected FilterBlock getNextBlock() {
-    int resolution = _resolution.getLowestResolution();
-    long h3Id = _h3Core.geoToH3(_geometry.getCoordinate().x, _geometry.getCoordinate().y, resolution);
-    assert _h3IndexReader != null;
+    if (_upperBound < 0 || _lowerBound > _upperBound) {
+      // Invalid upper bound
 
-    // find the number of rings based on distance for full match
-    // use the edge of the hexagon to determine the rings are within the distance. This is calculated by (1) divide the
-    // distance by edge length of the resolution to get the number of contained rings (2) use the (floor of number - 1)
-    // for fetching the rings since ring0 is the original hexagon
-    double edgeLength = _h3Core.edgeLength(resolution, LengthUnit.m);
-    int numFullMatchedRings = (int) Math.floor(_distance / edgeLength);
-    MutableRoaringBitmap fullMatchedDocIds = new MutableRoaringBitmap();
-    List<Long> fullMatchRings = new ArrayList<>();
-    if (numFullMatchedRings > 0) {
-      fullMatchRings = _h3Core.kRing(h3Id, numFullMatchedRings - 1);
-      for (long id : fullMatchRings) {
-        ImmutableRoaringBitmap docIds = _h3IndexReader.getDocIds(id);
-        fullMatchedDocIds.or(docIds);
+      return new FilterBlock(EmptyDocIdSet.getInstance());
+    }
+
+    if (Double.isNaN(_lowerBound) || _lowerBound < 0) {
+      // No lower bound
+
+      if (Double.isNaN(_upperBound)) {
+        // No upper bound
+        return new FilterBlock(new MatchAllDocIdSet(_numDocs));
```

Comment:
shall we have some tests for this algorithm?

Reply:
Added